rfcs/0004-transaction.md

```diff
@@ -0,0 +1,291 @@
+# SlateDB Transaction
+
+<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->
+
+- [SlateDB Transaction](#slatedb-transaction)
+  - [Background](#background)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Constraints](#constraints)
+  - [References](#references)
+  - [Proposal](#proposal)
+    - [API](#api)
+    - [Isolation Levels](#isolation-levels)
+    - [Prerequisites](#prerequisites)
+      - [Snapshot](#snapshot)
+      - [Write Batch](#write-batch)
+  - [Conflict Checking: Snapshot Isolation](#conflict-checking-snapshot-isolation)
+  - [Conflict Checking: Serializable Snapshot Isolation](#conflict-checking-serializable-snapshot-isolation)
+  - [Roadmap](#roadmap)
+  - [Updates](#updates)
+
+<!-- TOC end -->
+
+Status: Reviewing
+
+Authors:
+
+* [Li Yazhou](https://github.com/flaneur2020)
+
+## Background
```

Comment:
This RFC provides a lot of background context, with relevant references and examples. They are indeed very useful to understand the proposal.

Until I read till the end, it was harder for me to follow. 
The skeleton used in [compaction](https://github.com/slatedb/slatedb/pull/260/files) and other RFCs feels easier to read.

I think this RFC is written from the perspective of how the design was explored, rather than presenting the final design after the exploration. Some changes could  improve longer term readability. I am sharing few potentials.

1. Goals. I see the two isolation levels explained later. Listing them as goals, and providing simple use case examples would be faster to understand.
2. How other systems like RocksDB and Badger does isolation is interleaved in the design section, it would flow better if they are separated. i.e. list the design for slate db in a separate section and have a related systems as a separate section.